subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/ServiceReferencePropertyAnnotationHandler.java

```diff
@@ -0,0 +1,90 @@
+import static org.gradle.internal.reflect.validation.Severity.ERROR;
+
+public class ServiceReferencePropertyAnnotationHandler implements PropertyAnnotationHandler {
+    @Override
+    public Class<? extends Annotation> getAnnotationType() {
+        return ServiceReference.class;
+    }
+
+    @Override
+    public ImmutableSet<? extends AnnotationCategory> getAllowedModifiers() {
+        return ImmutableSet.of(OPTIONAL);
+    }
+
+    @Override
+    public boolean isPropertyRelevant() {
+        return true;
+    }
+
+    @Override
+    public boolean shouldVisit(PropertyVisitor visitor) {
+        return true;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void visitPropertyValue(String propertyName, PropertyValue value, PropertyMetadata propertyMetadata, PropertyVisitor visitor, BeanPropertyContext context) {
+        if (propertyMetadata.isAnnotationPresent(ServiceReference.class)) {
+            ServiceReference annotation = (ServiceReference) propertyMetadata.getAnnotationForCategory(AnnotationCategory.TYPE);
+            String serviceName = StringUtils.trimToNull(annotation.value());
+            visitor.visitServiceReference(propertyName, propertyMetadata.isAnnotationPresent(Optional.class), value, serviceName);
+        }
+    }
+
+    @Override
+    public void validatePropertyMetadata(PropertyMetadata propertyMetadata, TypeValidationContext validationContext) {
+        Method getter = propertyMetadata.getGetterMethod();
+        ModelType propertyType = ModelType.returnType(getter);
+        List<ModelType<?>> typeVariables = Cast.uncheckedNonnullCast(propertyType.getTypeVariables());
+        if (typeVariables.size() != 1 || !BuildService.class.isAssignableFrom(typeVariables.get(0).getRawClass())) {
+            validationContext.visitPropertyProblem(problem ->
+                problem.withId(ValidationProblemId.SERVICE_REFERENCE_MUST_BE_A_BUILD_SERVICE)
+                    .forProperty(propertyMetadata.getPropertyName())
+                    .reportAs(ERROR)
+                    .withDescription(() -> String.format("has @ServiceReference annotation used on property of type '%s' which is not a build service implementation", typeVariables.get(0).getName()))
+                    .happensBecause(() -> String.format("A property annotated with @ServiceReference must be of a type that implements '%s'", BuildService.class.getName()))
+                    .addPossibleSolution(String.format("Make '%s' implement '%s'", typeVariables.get(0).getName(), BuildService.class.getName()))
+                    .addPossibleSolution(String.format("Replace the @ServiceReference annotation on '%s' with @Internal and assign a value of type '%s' explicitly", propertyMetadata.getPropertyName(), typeVariables.get(0).getName()))
+                    .documentedAt("validation_problems", "service_reference_must_be_a_build_service")
+            );
+            return;
```

Comment:
The intent was to avoid multiple errors being reported once new cases were added, however 1) I guess we could want to report multiple errors, and 2) nowhere else do we do this, so removing...